SwiftGenerics.playground/Contents.swift

```diff
@@ -0,0 +1,113 @@
+import UIKit
+
+struct CountedSet<Element: Hashable> {
+    private(set) var memberDict: [Element: Int] = [:]
+    
+    mutating func insert(_ member: Element, occurrences: Int = 1) -> Int {
+        if occurrences > 0 {
+            if let count = memberDict[member] {
+                memberDict[member] = count + occurrences
+                return count + occurrences
+            } else {
+                memberDict[member] = occurrences
+                return occurrences
+            }
+        }
+        else { return 0 }
+    }
+    
+    mutating func remove(_ member: Element, occurrences: Int = 1) -> Int {
+        if let count = memberDict[member],
+            count >= occurrences {
+            if count > occurrences {
+                memberDict[member] = count - occurrences
+                return count - occurrences
+            } else {
+                memberDict[member] = nil
+                return 0
+            }
+        } else {
+            return 0
+        }
+    }
+    
+    subscript(_ member: Element) -> Int {
+        get {
+            if let count = memberDict[member] {
+                return count
+            } else {
+                return 0
+            }
+        }
+    }
+    
+    var count: Int {
+        return memberDict.count
+    }
+    var isEmpty: Bool {
+        return memberDict.isEmpty
+    }
+
+}
+
+var animals = CountedSet(memberDict: ["dog": 1])
+animals.insert("cat")
+animals.insert("rabbit")
+animals.insert("dog")
+animals.insert("cat")
+animals.remove("rabbit")
+animals.remove("dog")
+animals.count
+animals["dog"]
+animals["cat"]
+animals["rabbit"]
+
+
+extension CountedSet: ExpressibleByArrayLiteral {
+
+    init(arrayLiteral: Element...) {
+        for element in arrayLiteral {
+            insert(element)
+        }
+    }
+    
+    init<S: Sequence>(_ sequence: S) where
+      S.Iterator.Element == Element {
+      for element in sequence {
+        insert(element)
+      }
+    }
+
+    init<S: Sequence>(_ sequence: S) where
+      S.Iterator.Element == (key: Element, value: Int) {
+      for (element, count) in sequence {
+        insert(element, occurrences: count)
+      }
+    }
```

Comment:
Nice addition! So the top one would work for things like arrays and sets, and the second one would work for initializing with a dictionary.